### PR TITLE
[Perf] Fuse DeepEncoder relative bias in Triton attention

### PR DESCRIPTION
**Pull request by Levius-Fubuki**
## Purpose

DeepEncoder relative attention currently expands decomposed height and width
terms into a dense `[B, heads, HW, HW]` bias before SDPA. Its 64x64 global
attention therefore materializes a 384 MiB BF16 temporary.

This PR adds a dedicated Triton forward kernel that evaluates both relative
terms while streaming key/value tiles through online softmax:

```text
before: rel_h + rel_w -> dense 4D bias -> SDPA
after:  compact rel_h/rel_w -> fused Triton attention
```

The Triton path is limited to the production DeepEncoder global-attention
configuration on CUDA: FP16/BF16 with head dimension 64. Window attention,
CPU, unsupported dtypes, and unsupported head dimensions retain the existing
SDPA path. The implementation does not use FlexAttention or `torch.compile`.

### Regression tests

```bash
PYTHONPATH="$PWD" .venv/bin/python -m pytest -q \
  tests/models/multimodal/test_deepencoder.py

source /etc/network_turbo
PYTHONPATH="$PWD" .venv/bin/python -m pytest -q \
  tests/models/multimodal/processing/test_deepseek_ocr.py
```

Results:

```text
16 passed
6 passed
```

The focused suite covers FP16/BF16 numerical equivalence on square and
non-square grids, partial tiles, CPU fallback, mixed-device input rejection,
global/window dispatch, exact benchmark repeat counts, and CUDA Graph
capture/replay at 14x14 and 64x64.

The two upstream DeepSeek-OCR cases in
`tests/models/multimodal/generation/test_vit_cudagraph.py` remain skipped by
their existing markers (`2 skipped, 26 deselected`).

### Changed-file checks

```bash
ruff check \
  vllm/model_executor/models/deepencoder.py \
  tests/models/multimodal/test_deepencoder.py \
  benchmarks/kernels/benchmark_deepencoder_rel_pos_attention.py

ruff format --check \
  vllm/model_executor/models/deepencoder.py \
  tests/models/multimodal/test_deepencoder.py \
  benchmarks/kernels/benchmark_deepencoder_rel_pos_attention.py
```

`ruff-check`, `ruff-format`, `typos`, Python bytecode compilation, and
`git diff --check` pass. The full pre-commit wrapper could not finish its first
actionlint environment installation on the test server; the Python hooks that
apply to these files were run directly from the pinned hook environments.

### RTX 4090 microbenchmark

```bash
PYTHONPATH="$PWD" .venv/bin/python \
  benchmarks/kernels/benchmark_deepencoder_rel_pos_attention.py \
  --implementation production \
  --shapes global \
  --warmup 10 --repeats 50 \
  --output production-global.json
```

Hardware/software: NVIDIA RTX 4090, CUDA 13.2, PyTorch 2.13.0+cu132,
Triton 3.7.1, BF16. Results are medians of 50 interleaved samples.

| Scope | Path | p50 (ms) | Incremental peak allocation (MiB) | Max abs error |
|:---|:---|---:|---:|---:|
| attention operator | Dense SDPA | 1.1453 | 390 | reference |
| attention operator | Triton | 0.5478 | 6 | 9.77e-4 |
| full `RelPosAttention` | Dense SDPA | 1.5667 | 426 | reference |
| full `RelPosAttention` | Triton | 0.9820 | 42 | 4.88e-4 |

The independent recheck reduced operator latency by 52.2% and full-layer
latency by 37.3%, while both measurements saved exactly 384 MiB of incremental
peak allocation. Reference and candidate both pass CUDA Graph capture/replay.

The previous FlexAttention candidate measured 1.4469 ms for the full layer on
the same GPU. The Triton revision is about 32.1% lower than that result and has
no per-process `torch.compile` dependency.

### DeepSeek-OCR end-to-end correctness

I ran `deepseek-ai/DeepSeek-OCR` with a deterministic 1024x1024 OCR image,
greedy decoding, and `enforce_eager=True`. The base reference and Triton
candidate generated exactly the same 21 token IDs and decoded text.

## Limitations

These are isolated DeepEncoder attention results from one RTX 4090, not an
end-to-end serving throughput claim or a performance claim for other GPU
architectures. The optimized path is forward-only and intentionally limited to
CUDA FP16/BF16 with head dimension 64; other configurations use SDPA.

## AI assistance

AI assistance was used during implementation, testing, benchmarking, and PR
drafting. The human contributor reviewed and takes responsibility for the
change and evidence.

### CODE REVIEW

File: benchmarks/kernels/benchmark_deepencoder_rel_pos_attention.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+    expected = reference()
+    torch.cuda.synchronize()
+    compile_start = time.perf_counter()
+    actual = candidate()
+    torch.cuda.synchronize()
+    compile_seconds = time.perf_counter() - compile_start
+    correctness = errors(actual, expected)
+
+    rounds = 5
+    per_round = max(1, repeats // rounds)
```
[the remark this answers is not in the document]
Comment (Levius-Fubuki) [reply benchmarks/kernels/benchmark_deepencoder_rel_pos_attention.py]: Fixed. The benchmark now distributes every requested repeat across up to five alternating rounds and rejects non-positive values. Added coverage for 1, 5, 7, 51, 0, and -1; the focused suite passes with 12 tests, and an actual `--repeats 7` run records exactly 7 samples for each path.

File: vllm/model_executor/models/deepencoder.py [hunk 2]
```diff
@@ -264,6 +267,28 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return x
 
 
+_flex_attention_compiled = torch.compile(flex_attention, fullgraph=True)
```
Comment (Isotr0py) [inline vllm/model_executor/models/deepencoder.py]: Actually, we're deprecating torch.compile gradually. 😅

Comment (Levius-Fubuki) [reply vllm/model_executor/models/deepencoder.py]: Understood. I take this to mean that new optimization paths should avoid introducing `torch.compile`, even locally. Since FlexAttention relies on compilation for good performance, would you prefer a dedicated Triton implementation that fuses DeepEncoder global attention with the relative position bias? I’d like to confirm that direction before reworking the implementation.

Comment (Isotr0py) [reply vllm/model_executor/models/deepencoder.py]: Sounds great

File: vllm/model_executor/kernels/deepencoder_attention.py
Comment (Isotr0py) [inline vllm/model_executor/kernels/deepencoder_attention.py:1]: Since only deepencoder uses this kernel, let's put it at model implementation instead.

Comment (Levius-Fubuki) [reply vllm/model_executor/kernels/deepencoder_attention.py:1]: Moved it into `deepencoder.py` since it’s only used by DeepEncoder. I also tightened the launcher check so all five inputs have to be CUDA tensors on the same device.

File: vllm/model_executor/kernels/deepencoder_attention.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    k: torch.Tensor,
+    v: torch.Tensor,
+    rel_h: torch.Tensor,
+    rel_w: torch.Tensor,
+    height: int,
+    width: int,
+    scale: float,
+) -> torch.Tensor:
+    """Run DeepEncoder global attention with fused relative-position bias."""
+    assert q.is_cuda and k.is_cuda and v.is_cuda
```
[the remark this answers is not in the document]
Comment (Levius-Fubuki) [reply vllm/model_executor/kernels/deepencoder_attention.py]: Fixed. The launcher now validates `q`, `k`, `v`, `rel_h`, and `rel_w` before the Triton launch. Added a mixed-device regression test that reproduces the previous low-level driver failure and now verifies early rejection.

**Review by Isotr0py — APPROVED** [review]
